pkg/controller/core/workload_controller.go

```diff
@@ -324,6 +338,29 @@ func isDisabledRequeuedByReason(w *kueue.Workload, reason string) bool {
 	return cond != nil && cond.Status == metav1.ConditionFalse && cond.Reason == reason
 }
 
+// reconcileMaxExecutionTime deactivates the workload if its MaximumExecutionTime exceeded or returns a retry after value.
+func (r *WorkloadReconciler) reconcileMaxExecutionTime(ctx context.Context, wl *kueue.Workload) (time.Duration, error) {
+	admittedCondition := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted)
+	if admittedCondition == nil || admittedCondition.Status != metav1.ConditionTrue || wl.Spec.MaximumExecutionTimeSeconds == nil {
+		return 0, nil
+	}
+
+	remainingTime := time.Duration(*wl.Spec.MaximumExecutionTimeSeconds)*time.Second - r.clock.Since(admittedCondition.LastTransitionTime.Time)
+	if remainingTime > 0 {
+		return remainingTime, nil
+	}
+
+	if ptr.Deref(wl.Spec.Active, true) {
+		wl.Spec.Active = ptr.To(false)
```

Comment:
```suggestion
```

Could we execute the deactivation at https://github.com/kubernetes-sigs/kueue/blob/7be3b1e65ae1f19410e1b04bb7638f3fb543d688/pkg/controller/core/workload_controller.go#L154 in the next reconciliation.

Atcually, the `worklod.ApplyAdmissionStatus` performers only status update and this active spec update is not performed: https://github.com/epam/kubernetes-kueue/blob/5a684d22e3a82dff2d4f15b0c9a47d61d9e05a9c/pkg/workload/workload.go#L328

Reply:
good catch , it was a leftover from the initial "direct deactivation"  implementation.